### DISCUSSION TITLE
is register_uri_scheme_protocol routing with IPC ?

### DISCUSSION THREAD

**Discussion by ggslayer** [tauri-apps/tauri#8571] [Q&A] [answered]
Hi, I want to know is register_uri_scheme_protocol/register_async_uri_scheme_protocol  routing with IPC?   is there different in distribution version and debug version,  for example, debug version is http debug by web server? 

Additional, If I already register a protocol with name "myapp",  when I open a window with windowUrl parameter,  can I pass the "myapp://localhost/sub1/index.html"   or "myapp://sub1/index.html" ?   

thank you

**Comment by FabianLars** [comment]
> I want to know is register_uri_scheme_protocol/register_async_uri_scheme_protocol routing with IPC?

It should be, at least with the https://tauri.app/v1/api/config/#securityconfig.dangerousremotedomainipcaccess config (exists in v2 too). If it does not it's a bug we need to fix.

> is there different in distribution version and debug version, for example, debug version is http debug by web server?

This will mostly depend on your, but generally no. debug and production versions will both use custom protocol with the same implementation behind the scenes. If you want it to use a http server you'll have to do that yourself (redirect, or a conditional window url config).

> Additional, If I already register a protocol with name "myapp", when I open a window with windowUrl parameter, can I pass the "myapp://localhost/sub1/index.html" or "myapp://sub1/index.html" ?

Probably the first one, but just try it out, maybe both work.

**Reply by ggslayer** [reply]
thanks, I want to use IPC, because I think there is no socket port expose, It's cleaner than http,   I want to know futher more, 
1. does IPC's performance is better than http?
2. does IPC's safe is better than http? 

If yes, In order to achieve the performance and safe goal, what should I do more?
for example: config ...  and, want config should I research detail?

**Reply by FabianLars** [reply, marked as answer]
> does IPC's performance is better than http?

No, in v1 it's _much_ slower than a http server. In v2 it's much faster compared to v1 but we didn't measure it yet so i can't tell how fast it is compared to a http server (should be relatively close though)

> does IPC's safe is better than http?

Yes. As you said, not port etc. Only the webview itself can access the IPC as opposed to the whole system when using an http server.

> If yes, In order to achieve the performance and safe goal, what should I do more?
> for example: config ... and, want config should I research detail?

Can't think of anything general here. The ipc and custom protocols don't have much options config-wise so it should be mostly ready out of the box unless i'm forgetting something.

**Comment by ggslayer** [comment]
ok thankyou



---- Replied Message ----
| From | ***@***.***> |
| Date | 01/09/2024 20:04 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [tauri-apps/tauri] is register_uri_scheme_protocol routing with IPC ? (Discussion #8571) |

does IPC's performance is better than http?

No, in v1 it's much slower than a http server. In v2 it's much faster compared to v1 but we didn't measure it yet so i can't tell how fast it is compared to a http server (should be relatively close though)

does IPC's safe is better than http?

Yes. As you said, not port etc. Only the webview itself can access the IPC as opposed to the whole system when using an http server.

If yes, In order to achieve the performance and safe goal, what should I do more?
for example: config ... and, want config should I research detail?

Can't think of anything general here. The ipc and custom protocols don't have much options config-wise so it should be mostly ready out of the box unless i'm forgetting something.

**Reply by ggslayer** [reply]
@FabianLars  Hi, please give me some help.
I start a file web server in alpha v2,  which serve my local file dir.  In the dir,  deploy my website app with vue&tauri js sdk, which visit tauri commands in rust. 

I modify tauri.conf.json, add security field, config:
```
{
  "build": {
    "beforeDevCommand": "npm run dev",
    "beforeBuildCommand": "npm run build",
    "devPath": "http://localhost:1420",
    "distDir": "../dist"
  },
  "package": {
    "productName": "assistor",
    "version": "0.0.0"
  },
  "tauri": {
    "bundle": {
      "active": true,
      "targets": "all",
      "identifier": "com.assistor.dev",
      "icon": [
        "icons/32x32.png",
        "icons/128x128.png",
        "icons/128x128@2x.png",
        "icons/icon.icns",
        "icons/icon.ico"
      ]
    },
    "security": {
      "csp": null,
      "dangerousRemoteDomainIpcAccess": [
        {
          "domain": "localhost",
          "windows": [
            "main"
          ]
        }
      ]
    }
  },
  "plugins": {
    "shell": {
      "open": true
    }
  }
}
```
I build a window in rust, navigit it to http://localhost:5800/index.html (5800 is my web server listening), the html & js is loaded normally, when I click the button to call command in rust,  exception ocurred: "Not allowed by the scope".

question:
Is my security config is correct?  Is it enough?
when I add "dangerousUseHttpScheme" config item, It could't build succeed,  Is the option"dangerousUseHttpScheme" needed?
any other I should to config?  Is there any example or document descript the scene I want to archive?  (use tauri window load extenal page, and in the page, call tauri commands normally)

**Reply by FabianLars** [reply]
>  Is the option"dangerousUseHttpScheme" needed?

No. This is only applied to tauri's default protocol it serves your distDir from.

> exception ocurred: "Not allowed by the scope".

Can you try `"domain": "localhost:5800"`? I think the port is required 🤔

**Reply by ggslayer** [reply]
I'll try it tonight, after that, feedback here, thank you.
If ok, I want to know, Is there some method can modify the "domain" programmally?  because the port is picker dyamiclly in user's machine, not a fixed port.

other question:
in the dangerousRemoteDomainIpcAccess config, is windows must be given?  how to allow all windows in app access ipc?

**Reply by FabianLars** [reply]
> If ok, I want to know, Is there some method can modify the "domain" programmally? because the port is picker dyamiclly in user's machine, not a fixed port.

Yes, but only in rust (doing it in js would defeat the purpose), we only have this small example i think https://docs.rs/tauri/2.0.0-alpha.20/tauri/scope/ipc/struct.Scope.html#method.configure_remote_access

> in the dangerousRemoteDomainIpcAccess config, is windows must be given? how to allow all windows in app access ipc?

Right now specific window labels are required, there's an open feature request for regex/prefix support here though https://github.com/tauri-apps/tauri/issues/7461

**Reply by FabianLars** [reply]
Also, there will be a major allowlist/scope refactor before 2.0 hits beta so what i said may change again.

**Reply by ggslayer** [reply]
Hi, @FabianLars,   after effort, all goes ok,  I found:
1. there is no need port after localhost.
2. debug in the tauri code, found there is extra plugin check, So, after plugin permition,  all ok. (the tauri api I called is plugin process's api )

thank you!  I will continue my work
